master/static/migrations/20240212130408_proto-time-helper.tx.up.sql

```diff
@@ -0,0 +1,12 @@
+CREATE OR REPLACE FUNCTION public.proto_time(ts timestamptz)
+    RETURNS json
+    LANGUAGE sql IMMUTABLE
+    RETURNS NULL ON NULL INPUT
+AS $$
+    SELECT json_build_object(
+        -- Seconds since epoch
+        'seconds',  floor(extract(epoch FROM ts))::bigint, 
+        -- Fractional part in nanos since epoch
+        'nanos',    (MOD(extract(milliseconds FROM ts)::decimal, 1000::decimal)*1000000)::int
```

Comment:
why not just `extract(microseconds FROM ts)::int * 10e3`

Reply:
i'm not quite getting what the mod is supposed to do.

Reply:
ah, duh, wait. millis includes the seconds. i get it, mod'ing away the seconds.

Reply:
ok lgtm, though, once i realized `extract(millis..` includes more information than i thought. dang it is frustrating that there isn't a more straightfoward way to do this. i feel like this is slightly more obvious, but maybe only slightly `extract(seconds FROM ts)::int * 10e9 - extract(microseconds FROM ts)::int * 10e3` or something.

Reply:
I tried `extract(microseconds FROM ts)::int * 10e3 - extract(seconds FROM ts)::int * 10e9 ` and I'm getting a negative number haha

Reply:
weird.
```
with q as (select now() as ts)     
select extract(microseconds FROM ts)::int * 10e3 - extract(seconds FROM ts)::int * 10e9 from q;
```
works for me.